Campaign import export user docs (#361)
mautic/user-documentation: docs/campaigns/importing_campaigns.rst

Importing Campaigns
###################

The import feature allows you to add pre-configured Campaigns to your Mautic instance using ZIP files that contain all the relevant data needed to construct a Campaign.

.. vale off

Importing a Campaign
********************

.. vale on

You can import a Campaign in three ways.

1.\ Using Mautic instance
=========================

#. Click on the **Campaigns** menu to view existing Campaigns.

#. Click the **Options** button that resembles a cog icon in the top-right corner.

#. Select **Import** from the dropdown menu.

   |

   .. image:: images/import_campaign.png
      :alt: Highlight of Campaign menu, Options button, and Import option in the Mautic Campaigns section

   |

#. Choose the Campaign ZIP file\* you wish to import, then click the **Upload** button.

   |

   .. image:: images/upload_zip_file.png
      :alt: The Choose file and Upload buttons to import Campaigns

   |

   .. tip::

      \* Use a ZIP file created from the Mautic export function - **recommended**. This ensures inclusion of Campaign data, external Assets, and Dynamic Content.

> #. Select **Actions** option from the dropdown menu for the **Campaign** and **Segment** entities. Choose either **Update entity** or **Create new entity**.

Reviewer: ```suggestion
#. Select **Actions** option from the dropdown menu for the **Campaign** and **Segment** entities. Choose either **Update entity** or **Create new entity**.

   These options appear when the import process detects existing Campaigns or Segments that match the imported data.

   * **Update entity** - Overwrites the existing Campaign or Segment in Mautic with the imported data. Use this option to update configuration, but note that it can overwrite or destroy existing changes on your instance.
   * **Create new entity** - Forces Mautic to create a brand-new Campaign or Segment, even if a matching one already exists. This is the safer option to prevent data loss, but it can lead to duplicate resources.
```

Reviewer 2: Descriptions are accurate and match Ruth's request. Optional: a short lead-in saying these options appear because the import found a matching entity would make the "why you see it" part explicit. Otherwise looks good.

Reviewer: Does the short lead-in here clear enough?

Reviewer 2: Yes, "These options appear when the import process detects existing Campaigns or Segments that match the imported data" nails the "why am I seeing this" part. Reads well  good to go. 👍

#. Click the **Proceed** button.

   |

   .. image:: images/select_actions_import_campaigns.png
      :alt: Highlight of Actions dropdown menu and Proceed buttons in Mautic Import Campaigns 

   |

Once the import is successful, you should see a success notification.

.. image:: images/campaign_import_successful.png
   :alt: Campaign import successful notification

.. important::

   The Mautic instance only supports importing ZIP files. You can use both the command line and API endpoints to import correctly structured JSON files.

.. vale off

Activating an imported Campaign
-------------------------------

.. vale on

Follow the steps below to activate an imported Campaign:
      
#. Click on the **Campaigns** menu.
#. Locate the newly imported Campaign.
#. Click the red toggle button next to the Campaign's name to change the status to active.

   |

   .. image:: images/campaign_inactive_toggle_button.png
      :alt: Highlight of a Campaign's inactive toggle button

   |

> #. Click **Yes** when a prompt message appears.

Reviewer: ```suggestion
#. Click **Yes** when a prompt message appears.

   .. warning::
          
   Activating a Campaign starts execution immediately. Review all Campaign steps, configurations, and associated Assets thoroughly before turning on a Campaign to prevent unintended execution of events. 
          
   When a Campaign goes live, scheduled events with relative delays follow the **Campaign Republish Behavior** configuration. If the Campaign uses a specific setting instead of the global default, the execution behavior changes:

   * **Use global setting** - Applies the default behavior configured in the global Mautic settings.
   * **Restart on republish** - Resets the delay timer completely. The delay period starts over from zero the moment the Campaign becomes active.
   * **Count delay only while published** - Pauses the delay timer while the Campaign is inactive. The timer resumes from where it paused once the Campaign becomes active again.
   * **Count delay regardless of publish state** - default option. Keeps the delay timer running continuously in the background, even while the Campaign remains inactive.
```

Reviewer 2: The warning addresses Ruth's point well. Two fixes on the option list:

  1. Typo: "Count delay on while published" - should be "Count delay only while published" (the description is correct). 
  2. "Count delay regardless of publish state" is the default option - worth stating that explicitly.

Non-blocking: the Republish Behavior detail applies to republishing in general rather than to import specifically it could live in a note or a link to keep the import step focused.

Reviewer: > Non-blocking: the Republish Behavior detail applies to republishing in general rather than to import specifically it could live in a note or a link to keep the import step focused.

I think that's a great idea as I don't think we have a dedicated section or whatsoever in the docs for this info. I'll create an issue to address this. 👍🏼

Reviewer 2: Sounds good separate issue for the Republish Behavior section makes sense. We can keep the warning here focused on activation, then link out once the dedicated  section exists.

Reviewer: @vukovicpredrag quick questions:

- Does Republish behavior only used in Campaign? If not, where else?
- Do you happen to know when did this first introduce? M7 or M5? I need the info for backport matter.

TIA!

          # Add other campaign details as needed
      }

      # Send import request
      response = requests.post(url, headers=headers, json=payload)

      # Handle response
      if response.status_code == 200:
          imported_campaign = response.json()
          print("Campaign imported successfully")

API import methods
------------------

Mautic supports two primary methods of API-based Campaign import:

.. vale off

#. **ZIP file import**
   
   * Use ``multipart/form-data`` content type
   * Upload the complete Campaign export ZIP file
   * Includes all Campaign Assets and dependencies from the ZIP file

#. **JSON data import**
   
   * Use ``application/json`` content type
   * Send Campaign details directly in the request body
   * Useful for creating new Campaigns or updating existing Campaigns

.. vale on

.. important::
    
   * Replace ``example.com`` with your actual Mautic instance domain.
   * Ensure you have a valid access token by accessing the API Credentials section within Mautic's settings.
   * The imported Campaign must comply with Mautic's Campaign structure.
   * Verify import permissions and data integrity.

.. vale off

How Campaign import works

> ********************************

Reviewer: ```suggestion
How Campaign import works
*************************
```

Importing Campaigns
###################

The import feature allows you to add pre-configured Campaigns to your Mautic instance using ZIP files that contain all the relevant data needed to construct a Campaign.

.. vale off

Importing a Campaign
********************

.. vale on

You can import a Campaign in three ways.

1.\ Using Mautic instance
=========================

#. Click on the **Campaigns** menu to view existing Campaigns.

#. Click the **Options** button that resembles a cog icon in the top-right corner.

#. Select **Import** from the dropdown menu.

   |

   .. image:: images/import_campaign.png
      :alt: Highlight of Campaign menu, Options button, and Import option in the Mautic Campaigns section

   |

#. Choose the Campaign ZIP file\* you wish to import, then click the **Upload** button.

   |

   .. image:: images/upload_zip_file.png
      :alt: The Choose file and Upload buttons to import Campaigns

   |

   .. tip::

      \* Use a ZIP file created from the Mautic export function - **recommended**

#. Ensures inclusion of Campaign data, external Assets, and Dynamic Content.

> #. Select **Actions** option from the dropdown menu for the **Campaign** and **Segment** entities. Choose either **Update entity** or **Create new entity**.

Reviewer 3: I think we should explain what each does and why you see it - e.g. update entity suggests that it has found this already existing in Mautic, and it'll update with any changes compared to what's there (could be destructive, might overwrite changes). Create new will force it to create a new entity, even if something that matches is found. Safer option but can lead to duplication of resources.

#. Click the **Proceed** button.

   |

   .. image:: images/select_actions_import_campaigns.png
      :alt: Highlight of Actions dropdown menu and Proceed buttons in Mautic Import Campaigns 

   |

Once the import is successful, you should see a success notification.

.. image:: images/campaign_import_successful.png
   :alt: Campaign import successful notification

.. important::

   The Mautic instance only supports importing ZIP files. You can use both the command line and API endpoints to import correctly structured JSON files.

.. vale off

Activating an imported Campaign
-------------------------------

.. vale on

Follow the steps below to activate an imported Campaign:
      
#. Click on the **Campaigns** menu.
#. Locate the newly imported Campaign.
#. Click the red toggle button next to the Campaign's name to change the status to active.

   |

   .. image:: images/campaign_inactive_toggle_button.png
      :alt: Highlight of a Campaign's inactive toggle button

   |

> #. Click **Yes** when a prompt message appears.

Reviewer 3: We should link to the docs which explain what this means. Clicking yes can have consequences. Publishing a campaign without checking it first could also be disastrous. I think we need to add a clear warning here, that once you publish a campaign, it immediately starts running. People must check that they understand what the campaign is doing, and verify each step of the way, before they decide to turn it on. Red warning admonition called for, I think.

How the importing Campaign works
********************************

During the import process, Mautic performs a comprehensive analysis of the data:

Permissions
    Checks that the logged-in User has the correct permissions to import.

Entities
    Identifies required entities for the Campaign to function.

Plugin validation
    The import function verifies Plugin installation. It checks if a Campaign template depends on an external Plugin. If a required Plugin is missing, the import process halts and prompts the User to install the necessary Plugin before continuing.

Conflict resolution
    Validates for potential ID conflicts in imported entities. Where conflicts exist, Mautic provides options to:
    
    * Update existing entities, allowing Administrators to update existing Campaigns.
    * Create new entities using a new ID.

Automatic data mapping
    Mautic maps imported data to the correct locations and automatically creates any necessary dependent entities.

Campaign activation
    After a successful import, the Campaign remains inactive by default.

> .. vale off

Reviewer: Same comment here with the one in the `exporting_campaigns.rst` regarding moving this section down.

Reviewer 2: Same as my comment on the export page - move this after the practical import steps.

.. vale off

Importing a Campaign
********************

.. vale on

You can import a Campaign in three ways.

1.\ Using Mautic instance
=========================

#. Click on the **Campaigns** menu to view existing Campaigns.

#. Click the **Options** button that resembles a cog icon in the top-right corner.

#. Select **Import** from the dropdown menu.

   |

   .. image:: images/import_campaign.png
      :alt: Highlight of Campaign menu, Options button, and Import option in the Mautic Campaigns section

   |

#. Choose the Campaign ZIP file\* you wish to import, then click the **Upload** button.

   |

   .. image:: images/upload_zip_file.png
      :alt: The Choose file and Upload buttons to import Campaigns

   |

   .. tip::

      \* Use a ZIP file created from the Mautic export function - **recommended**. This ensures inclusion of Campaign data, external Assets, and Dynamic Content.
#. Select **Actions** option from the dropdown menu for the **Campaign** and **Segment** entities. Choose either **Update entity** or **Create new entity**.

   These options appear when the import process detects existing Campaigns or Segments that match the imported data.

>    * **Update entity** - Overwrites the existing Campaign or Segment in Mautic with the imported data. Use this option to update configuration, but note that it can overwrite or destroy existing changes on your instance.
>    * **Create new entity** - Forces Mautic to create a brand-new Campaign or Segment, even if a matching one already exists. This is the safer option to prevent data loss, but it can lead to duplicate resources.

Reviewer: The word 'lead' that Vale warns is a verb. So, we can safely wrap this in vale off/on statements.

```suggestion
.. vale off

   * **Update entity** - Overwrites the existing Campaign or Segment in Mautic with the imported data. Use this option to update configuration, but note that it can overwrite or destroy existing changes on your instance.
   * **Create new entity** - Forces Mautic to create a brand-new Campaign or Segment, even if a matching one already exists. This is the safer option to prevent data loss, but it can lead to duplicate resources.

.. vale on
```

   :alt: Campaign import successful notification

.. important::

   The Mautic instance only supports importing ZIP files. You can use both the command line and API endpoints to import correctly structured JSON files.

.. vale off

Activating an imported Campaign
-------------------------------

.. vale on

Follow the steps below to activate an imported Campaign:
      
#. Click on the **Campaigns** menu.
#. Locate the newly imported Campaign.
#. Click the red toggle button next to the Campaign's name to change the status to active.

   |

   .. image:: images/campaign_inactive_toggle_button.png
      :alt: Highlight of a Campaign's inactive toggle button

   |

#. Click **Yes** when a prompt message appears.

   .. warning::
          
   Activating a Campaign starts execution immediately. Review all Campaign steps, configurations, and associated Assets thoroughly before turning on a Campaign to prevent unintended execution of events. 
          
   When a Campaign goes live, scheduled events with relative delays follow the **Campaign Republish Behavior** configuration. If the Campaign uses a specific setting instead of the global default, the execution behavior changes:

>    * **Use global setting** - Applies the default behavior configured in the global Mautic settings.
>    * **Restart on republish** - Resets the delay timer completely. The delay period starts over from zero the moment the Campaign becomes active.
>    * **Count delay only while published** - Pauses the delay timer while the Campaign is inactive. The timer resumes from where it paused once the Campaign becomes active again.
>    * **Count delay regardless of publish state** - default option. Keeps the delay timer running continuously in the background, even while the Campaign remains inactive.

Reviewer: The word 'published' that Vale flags here are what appears on the UI. Let's wrap this in vale off/on statements.

```suggestion
   .. vale off
   
   * **Use global setting** - Applies the default behavior configured in the global Mautic settings.
   * **Restart on republish** - Resets the delay timer completely. The delay period starts over from zero the moment the Campaign becomes active.
   * **Count delay only while published** - Pauses the delay timer while the Campaign is inactive. The timer resumes from where it paused once the Campaign becomes active again.
   * **Count delay regardless of publish state** - default option. Keeps the delay timer running continuously in the background, even while the Campaign remains inactive.
   
   .. vale on
```

**Command parameters**

* ``--entity=campaign``: specifies the type of entity to import, in this case, ``campaign``.

* ``--file=/tmp/entity_data.zip``: the path to the ZIP file containing the Campaign data.

* ``--user=<user_id>``: the ID of the User performing the import.

.. important::

   * Ensure the ZIP file is a valid Mautic Campaign export
   * The specified User must have appropriate import permissions
   * Verify the path is correct before running the command

3.\ Using Mautic API
====================

You can import Campaigns programmatically using the Mautic API.

#. **cURL example with ZIP file**

   .. code-block:: bash

      curl -X POST 'https://example.com/api/campaigns/import' \
      -H 'Authorization: Bearer YOUR_ACCESS_TOKEN' \
      -H 'Content-Type: multipart/form-data' \
      -F 'file=@/path/to/campaign_export.zip'

#. **Python example with JSON data**

   .. code-block:: python

      import requests
      import json

      # API Endpoint
      url = 'https://example.com/api/campaigns/import'

      # Authentication
      headers = {
          'Authorization': 'Bearer YOUR_ACCESS_TOKEN',
          'Content-Type': 'application/json'
      }

>       # Campaign import data

Reviewer: Follow the comment in the suggestion.

```suggestion
      # Body must be the full campaign export structure (same as entity_data.json)
```

**Command parameters**

* ``--entity=campaign``: specifies the type of entity to import, in this case, ``campaign``.

* ``--file=/tmp/entity_data.zip``: the path to the ZIP file containing the Campaign data.

* ``--user=<user_id>``: the ID of the User performing the import.

.. important::

   * Ensure the ZIP file is a valid Mautic Campaign export
   * The specified User must have appropriate import permissions
   * Verify the path is correct before running the command

3.\ Using Mautic API
====================

You can import Campaigns programmatically using the Mautic API.

#. **cURL example with ZIP file**

   .. code-block:: bash

      curl -X POST 'https://example.com/api/campaigns/import' \
      -H 'Authorization: Bearer YOUR_ACCESS_TOKEN' \
      -H 'Content-Type: multipart/form-data' \
      -F 'file=@/path/to/campaign_export.zip'

#. **Python example with JSON data**

   .. code-block:: python

      import requests

      # API Endpoint
      url = 'https://example.com/api/campaigns/import'

      # Authentication
      headers = {
          'Authorization': 'Bearer YOUR_ACCESS_TOKEN',
          'Content-Type': 'application/json'
      }

      # Campaign import data

>       payload = {

Reviewer 2: This JSON import example won't work as written. `CampaignApiController::importCampaignAction` requires the request body to be the full campaign export structure (it  checks `isset($data[0]['campaign'])`). A flat `{ "name": ..., "description": ... }` payload fails that check and returns `400`. The endpoint also returns `201`, not  `200`.

The body must be the same JSON as `entity_data.json` from an export ZIP (or the response of `GET /api/campaigns/export/{id}`). 

Suggested fix:

  ```python

  import requests
  import json

  url = 'https://example.com/api/campaigns/import'

  headers = {
      'Authorization': 'Bearer YOUR_ACCESS_TOKEN',
      'Content-Type': 'application/json'
  }

  # Body must be the full campaign export structure (same as entity_data.json)
  with open('entity_data.json') as f:
      payload = json.load(f)
  
  response = requests.post(url, headers=headers, json=payload)

  if response.status_code == 201:
      print("Campaign imported successfully")
  ```

Reviewer: @vukovicpredrag I've applied the suggestion by copy-pasting the code block. I guess everything's good to go?

Reviewer 2: Sorry for the late reply @adiati98 , this slipped past me.  I've gone through the latest changes (including the Python example fix) and they all look correct  nothing further from my side, my approval stands. Thanks for your patience!

Reviewer: Thanks so much for your help and review, @vukovicpredrag!

>       import requests
> 
>       # API Endpoint
>       url = 'https://example.com/api/campaigns/import'
> 
>       # Authentication
>       headers = {
>           'Authorization': 'Bearer YOUR_ACCESS_TOKEN',
>           'Content-Type': 'application/json'
>       }
> 
>       # Campaign import data
>       payload = {
>           'name': 'Imported Campaign',
>           'description': 'Campaign imported via API',
>           # Add other campaign details as needed
>       }
> 
>       # Send import request
>       response = requests.post(url, headers=headers, json=payload)
> 
>       # Handle response
>       if response.status_code == 200:
>           imported_campaign = response.json()
>           print("Campaign imported successfully")

Reviewer: ```suggestion
   .. code-block:: python

      import requests
      import json

      # API Endpoint
      url = 'https://example.com/api/campaigns/import'

      # Authentication
      headers = {
          'Authorization': 'Bearer YOUR_ACCESS_TOKEN',
          'Content-Type': 'application/json'
      }

      # Campaign import data
      with open('entity_data.json') as f:
          payload = json.load(f)
      }

      # Send import request
      response = requests.post(url, headers=headers, json=payload)

      # Handle response
      if response.status_code == 201:
          print("Campaign imported successfully")
```
